### PR TITLE
Implement NoLeadingUnderscores rule and tests.

### PR DESCRIPTION
**Pull request by LaurenWhite**

### CODE REVIEW

File: tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift [hunk 1]
```diff
@@ -11,5 +11,104 @@ import SwiftSyntax
...
+    super.visit(node) // Visit children despite override
+  }
+  
+  public override func visit(_ node: EnumDeclSyntax) {
+    diagnoseUnderscoreViolation(name: node.identifier)
+    super.visit(node)
+  }
+  
+  public override func visit(_ node: EnumCaseDeclSyntax) {
+    var iterator = node.elements.makeIterator()
```
Comment (harlanhaskins) [inline tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift]: Nitpick: `for item in node.elements`

File: tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift [hunk 2]
```diff
@@ -11,5 +11,104 @@ import SwiftSyntax
...
+    while let parameter = paramIterator.next() {
+      if let typeIdentifier = parameter.firstName {
+        diagnoseUnderscoreViolation(name: typeIdentifier)
+      }
+      if let varIdentifier = parameter.secondName {
+        diagnoseUnderscoreViolation(name: varIdentifier)
+      }
+    }
+    // Check generic parameter names
+    var genParamIterator = node.genericParameterClause?.genericParameterList.makeIterator()
```
Comment (harlanhaskins) [inline tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift]: Same nitpick.

File: tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift [hunk 3]
```diff
@@ -11,5 +11,104 @@ import SwiftSyntax
...
+  
+  public override func visit(_ node: ProtocolDeclSyntax) {
+    diagnoseUnderscoreViolation(name: node.identifier)
+    super.visit(node)
+  }
+  
+  public override func visit(_ node: StructDeclSyntax) {
+    diagnoseUnderscoreViolation(name: node.identifier)
+    // Check generic parameter names
+    var iterator = node.genericParameterClause?.genericParameterList.makeIterator()
```
Comment (harlanhaskins) [inline tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift]: Same nitpick.

File: tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift [hunk 4]
```diff
@@ -11,5 +11,104 @@ import SwiftSyntax
...
+    super.visit(node)
+  }
+  
+  public override func visit(_ node: TypealiasDeclSyntax) {
+    diagnoseUnderscoreViolation(name: node.identifier)
+  }
+  
+  public override func visit(_ node: InitializerDeclSyntax) {
+    // Check parameter names of initializer
+    var iterator = node.parameters.parameterList.makeIterator()
```
Comment (harlanhaskins) [inline tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift]: Same nitpick.

**Comment by harlanhaskins** [comment]
Great work! Small nitpicks, but this is great.

File: tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift [hunk 5]
```diff
@@ -11,5 +11,104 @@ import SwiftSyntax
...
+    var iterator = node.elements.makeIterator()
+    while let item = iterator.next() {
+      diagnoseUnderscoreViolation(name: item.identifier)
+    }
+  }
+  
+  public override func visit(_ node: FunctionDeclSyntax) {
+    diagnoseUnderscoreViolation(name: node.identifier)
+    // Check parameter names of function
+    var paramIterator = node.signature.input.parameterList.makeIterator()
```
Comment (allevato) [inline tools/swift-format/Sources/Rules/NoLeadingUnderscores.swift]: Also use a `for` loop instead of an iterator here

**Comment by allevato** [comment]
This looks great! Merging it now.

**Pull request merged**